cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/ConstraintCheckerService.java

```diff
@@ -232,17 +233,21 @@ private void checkAndUpdateJob(JobQueueDataset jobQueue, Job job) {
           // - the transaction the marked it as to be deleted
           // - the subscriber's transaction that may not have seen that change
           (job.getState() == Job.State.PENDING_TRIGGER &&
-            System.currentTimeMillis() - job.getToBeDeletedTimestamp()
-              > 2 * Schedulers.SUBSCRIBER_TX_TIMEOUT_MILLIS))) {
+            now - job.getToBeDeletedTimestamp() > 2 * Schedulers.SUBSCRIBER_TX_TIMEOUT_MILLIS))) {
           jobQueue.deleteJob(job);
         }
         return;
       }
-      // TODO: check (now - jobCreationTime < job timeout), then delete it.
+      if (now - job.getCreationTime() >= job.getSchedule().getTimeoutMillis() +
```

Comment:
The timout computation involves `Schedulers.SUBSCRIBER_TX_TIMEOUT_MILLIS`, and I'd rather not push that into the dataset, so leaving the logic where it is currently.